src/validator/context.rs

```diff
@@ -0,0 +1,2246 @@
+    pub fn set_nsec3_iter_insecure(&mut self, value: u16) {
+        self.nsec3_iter_insecure = NSEC3_ITER_INSECURE.limit(value)
+    }
+
+    pub(crate) fn nsec3_iter_bogus(&self) -> u16 {
+        self.nsec3_iter_bogus
+    }
+
+    /// Set the number of NSEC3 iterations above which the result is
+    /// considered bogus.
+    ///
+    /// The value has to be at least zero, at most five hundred and the
+    /// default is one five hundred.
+    pub fn set_nsec3_iter_bogus(&mut self, value: u16) {
+        self.nsec3_iter_bogus = NSEC3_ITER_INSECURE.limit(value)
+    }
+
+    pub(crate) fn max_cname_dname(&self) -> u8 {
+        self.max_cname_dname
+    }
+
+    /// Set the maximum number of CNAME and DNAME records that are followed
+    /// during validation.
+    ///
+    /// The value has to be at least zero, at most one hundred and the
+    /// default is one eleven.
+    pub fn set_max_cname_dname(&mut self, value: u8) {
+        self.max_cname_dname = MAX_CNAME_DNAME.limit(value)
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            max_node_cache: MAX_NODE_CACHE.default(),
+            max_nsec3_cache: MAX_NSEC3_CACHE.default(),
+            max_isig_cache: MAX_ISIG_CACHE.default(),
+            max_usig_cache: MAX_USIG_CACHE.default(),
+            max_node_validity: MAX_NODE_VALIDITY.default(),
+            max_bogus_validity: MAX_BOGUS_VALIDITY.default(),
+            max_bad_signatures: MAX_BAD_SIGNATURES.default(),
+            nsec3_iter_insecure: NSEC3_ITER_INSECURE.default(),
+            nsec3_iter_bogus: NSEC3_ITER_BOGUS.default(),
+            max_cname_dname: MAX_CNAME_DNAME.default(),
+        }
+    }
+}
+
+//------------ ValidationContext ----------------------------------------------
+
+enum VGResult {
+    Groups(Vec<ValidatedGroup>),
+    Bogus(Option<ExtendedError<Vec<u8>>>),
+    Err(Error),
+}
+
+pub struct ValidationContext<Upstream> {
+    ta: TrustAnchors,
+    upstream: Upstream,
+    config: Config,
+
+    node_cache: Cache<Name<Bytes>, Arc<Node>>,
+    nsec3_cache: Nsec3Cache,
+    isig_cache: SigCache, // Signature cache for infrastructure.
+    usig_cache: SigCache, // Signature cache for user requests.
+}
+
+impl<Upstream> ValidationContext<Upstream> {
+    pub fn new(ta: TrustAnchors, upstream: Upstream) -> Self {
+        Self::with_config(ta, upstream, Default::default())
+    }
+
+    pub fn with_config(
+        ta: TrustAnchors,
+        upstream: Upstream,
+        config: Config,
+    ) -> Self {
+        Self {
+            ta,
+            upstream,
+            node_cache: Cache::new(config.max_node_cache),
+            nsec3_cache: Nsec3Cache::new(config.max_nsec3_cache),
+            isig_cache: SigCache::new(config.max_isig_cache),
+            usig_cache: SigCache::new(config.max_usig_cache),
+            config,
+        }
+    }
+
+    // On success, return the validation state and an optionally an extended DNS
+    // error.
```

Comment:
Is this more clear:
    /// Validate a DNS reply message. An Error value will be returned if the
    /// message cannot be parsed or if there is any other message-related.
    /// Otherwise the function return the DNSSEC validation state together
    /// with an optional EDNS(0) ExtendError. The ExtendedError, if present
    /// provides additional information on the cause of the validation state.
    /// This ExtendError is added to the reply message by the
    /// [validator](crate::net::client::validator) transport.

Reply:
Yes much better, although I think you're missing `error` after `message-related`.